File: web/src/subgroups.ts
import type { Lesson } from "./types";

export type SubgroupOption = { id: string; label: string };
export type SubgroupStream = { id: string; title: string; options: SubgroupOption[]; joined: boolean };
export type SubgroupMark = { streamId: string; options: SubgroupOption[]; chosenId: string | null; showChooser: boolean };

type Member = { streamId: string; optionIds: string[]; joined: boolean };
type Index = { streams: SubgroupStream[]; membership: Map<string, Member> };

export function subgroupIndex(lessons: Lesson[]): Index {
  const streams = new Map<string, { id: string; title: string; same: boolean; joined: boolean; options: Map<string, string> }>();
  const membership = new Map<string, Member>();
  for (const cluster of clusters(lessons)) {
    let builder = streams.get(cluster.streamId);
    if (!builder) {
      builder = { id: cluster.streamId, title: cluster.title, same: cluster.same, joined: true, options: new Map() };
      streams.set(cluster.streamId, builder);
    }
    builder.joined = builder.joined && cluster.lessons.length === 1;
    for (const [id, label] of cluster.labels) if (!builder.options.has(id)) builder.options.set(id, label);
    for (const lesson of cluster.lessons) {
      const key = lessonKey(lesson);
      const ids = cluster.same ? teacherNames(lesson.teacherRaw).map(teacherKey) : [slotOptionId(lesson)];
      const previous = membership.get(key);
      const joined = cluster.lessons.length === 1 && previous?.joined !== false;
      membership.set(key, { streamId: cluster.streamId, optionIds: ids, joined: joined && cluster.lessons.length === 1 });
    }
  }
  const built = [...streams.values()]
    .filter(builder => builder.options.size >= 2)
    .map(builder => ({
      id: builder.id,
      title: builder.title,
      joined: builder.joined && builder.same,
      options: [...builder.options].map(([id, label]) => ({ id, label })),
    }))
    .sort((a, b) => a.id.localeCompare(b.id));
  const live = new Set(built.map(stream => stream.id));
  for (const [key, member] of [...membership]) if (!live.has(member.streamId)) membership.delete(key);
  return { streams: built, membership };
}

export function visibleLessons(lessons: Lesson[], choices: Record<string, string>): Lesson[] {
  const index = subgroupIndex(lessons);
  return lessons.filter(lesson => keep(lesson, index, choices));
}

export function subgroupMark(lesson: Lesson, day: Lesson[], index: Index, choices: Record<string, string>): SubgroupMark | null {
  const member = index.membership.get(lessonKey(lesson));
  if (!member) return null;
  const stream = index.streams.find(item => item.id === member.streamId);
  if (!stream) return null;
  const chosen = choices[stream.id] && stream.options.some(option => option.id === choices[stream.id]) ? choices[stream.id] : null;
  const first = day.filter(item => index.membership.get(lessonKey(item))?.streamId === stream.id)
    .sort((a, b) => timeKey(a.timeStart).localeCompare(timeKey(b.timeStart)) || a.index - b.index || lessonKey(a).localeCompare(lessonKey(b)))[0];
  return { streamId: stream.id, options: stream.options, chosenId: chosen, showChooser: !!first && lessonKey(first) === lessonKey(lesson) };
}

function keep(lesson: Lesson, index: Index, choices: Record<string, string>): boolean {
  const member = index.membership.get(lessonKey(lesson));
  if (!member) return true;
  const stream = index.streams.find(item => item.id === member.streamId);
  if (!stream) return true;
  const choice = choices[stream.id];
  if (!choice || !stream.options.some(option => option.id === choice)) return true;
  if (member.joined) return true;
  return member.optionIds.includes(choice);
}

function clusters(lessons: Lesson[]) {
  const found: { streamId: string; title: string; same: boolean; lessons: Lesson[]; labels: Map<string, string> }[] = [];
  const seen = new Set<string>();
  const slots = new Map<string, Lesson[]>();
  for (const lesson of lessons) {
    const key = `${lesson.dayOfWeek}|${timeKey(lesson.timeStart)}`;
    const rows = slots.get(key) || [];
    rows.push(lesson);
    slots.set(key, rows);
  }
  for (const [slot, rows] of slots) {
    const [day, time] = slot.split("|");
    for (const week of weekCodes(rows)) {
      const active = rows.filter(lesson => lesson.parity === 0 || lesson.parity === week);
      const names = new Map<string, string>();
      for (const lesson of [...active].sort((a, b) => a.index - b.index || subjectKey(a).localeCompare(subjectKey(b)))) {
        for (const name of teacherNames(lesson.teacherRaw)) {
          const id = teacherKey(name);
          if (id && !names.has(id)) names.set(id, name.trim());
        }
      }
      if (names.size < 2 || active.length === 0) continue;
      const subjects = [...new Set(active.map(subjectKey).filter(Boolean))];
      const same = subjects.length === 1;
      const set = [...names.keys()].sort().join("+");
      const streamId = same ? `s:${subjects[0]}:${set}` : `t:${day}:${time}:${set}`;
      const signature = streamId + "#" + active.map(lessonKey).sort().join(",");
      if (seen.has(signature)) continue;
      seen.add(signature);
      const labels = same ? names : new Map(active.map(lesson => [slotOptionId(lesson), slotLabel(lesson)] as const));
      const title = [...active].sort((a, b) => a.dayOfWeek - b.dayOfWeek || timeKey(a.timeStart).localeCompare(timeKey(b.timeStart)) || a.index - b.index)[0].subjectRaw;
      found.push({ streamId, title, same, lessons: active, labels });
    }
  }
  return found;
}

function slotLabel(lesson: Lesson): string {
  const teachers = teacherNames(lesson.teacherRaw).join(", ");
  return [lesson.subjectRaw.trim(), teachers].filter(Boolean).join(" · ") || teachers;
}

function slotOptionId(lesson: Lesson): string {
  const teachers = teacherNames(lesson.teacherRaw).map(teacherKey).sort().join("+");
  return `${subjectKey(lesson)}|${teachers}|${(lesson.classroomRaw || "").trim()}`;
}

function lessonKey(lesson: Lesson): string {
  return [lesson.dayOfWeek, lesson.parity, lesson.index, timeKey(lesson.timeStart), subjectKey(lesson), teacherKey(lesson.teacherRaw || ""), (lesson.classroomRaw || "").trim()].join("|");
}

function subjectKey(lesson: Lesson): string {
  const norm = (lesson.subjectNormalized || "").trim();
  if (norm) return norm;
  return (lesson.subjectRaw || "").trim().toLocaleLowerCase("ru").replaceAll("ё", "е");
}

function teacherNames(raw: string | null | undefined): string[] {
  return (raw || "").split(";").map(part => part.trim()).filter(part => part && part !== "—" && part !== "-");
}

function teacherKey(name: string): string {
  return name.trim().toLocaleLowerCase("ru").replaceAll("ё", "е").replace(/[.\s]+/g, " ").trim();
}

function timeKey(raw: string): string {
  const match = /(\d{1,2}):(\d{2})/.exec(raw || "");
  if (!match) return (raw || "").trim();
  return `${match[1].padStart(2, "0")}:${match[2]}`;
}

function weekCodes(rows: Lesson[]): number[] {
  const codes = [...new Set(rows.map(lesson => lesson.parity).filter(parity => parity > 0))];
  return codes.length ? codes : [1];
}

File: android/app/src/main/java/ru/bgtu_voenmeh/zapara/data/Subgroups.kt
package ru.bgtu_voenmeh.zapara.data

import java.util.Locale

/**
 * A split is two or more teachers present in the same week at one bell.
 * Odd and even lessons at that bell are not a split: they never meet.
 */
object Subgroups {
    data class Option(val id: String, val label: String)
    data class Stream(val id: String, val title: String, val options: List<Option>, val joined: Boolean)
    data class Mark(val streamId: String, val options: List<Option>, val chosenId: String?, val showChooser: Boolean)

    class Index(
        val streams: List<Stream>,
        private val membership: Map<String, Member>
    ) {
        fun member(lesson: Lesson): Member? = membership[lessonKey(lesson)]
    }

    data class Member(val streamId: String, val optionIds: Set<String>, val joined: Boolean)

    fun index(lessons: List<Lesson>): Index {
        val clusters = clusters(lessons)
        val streams = linkedMapOf<String, StreamBuilder>()
        val membership = linkedMapOf<String, Member>()
        for (cluster in clusters) {
            val builder = streams.getOrPut(cluster.streamId) {
                StreamBuilder(cluster.streamId, cluster.title, cluster.sameSubject)
            }
            builder.joined = builder.joined && cluster.lessons.size == 1
            for ((id, label) in cluster.optionLabels) builder.options.putIfAbsent(id, label)
            for (lesson in cluster.lessons) {
                val key = lessonKey(lesson)
                val ids = if (cluster.sameSubject) teacherNames(lesson.teacherRaw).map { teacherKey(it) }.toSet()
                else setOf(slotOptionId(lesson))
                val previous = membership[key]
                val joined = cluster.lessons.size == 1 && (previous?.joined != false)
                membership[key] = Member(cluster.streamId, ids, joined && cluster.lessons.size == 1)
            }
        }
        return Index(
            streams.values.map { it.build() }.filter { it.options.size >= 2 }.sortedBy { it.id },
            membership.filterValues { member -> (streams[member.streamId]?.options?.size ?: 0) >= 2 }
        )
    }

    fun visible(lessons: List<Lesson>, choices: Map<String, String>): List<Lesson> {
        val built = index(lessons)
        return lessons.filter { keep(it, built, choices) }
    }

    fun keep(lesson: Lesson, index: Index, choices: Map<String, String>): Boolean {
        val member = index.member(lesson) ?: return true
        val stream = index.streams.firstOrNull { it.id == member.streamId } ?: return true
        val choice = choices[stream.id]?.takeIf { id -> stream.options.any { it.id == id } } ?: return true
        if (member.joined) return true
        return choice in member.optionIds
    }

    fun mark(lesson: Lesson, day: List<Lesson>, index: Index, choices: Map<String, String>): Mark? {
        val member = index.member(lesson) ?: return null
        val stream = index.streams.firstOrNull { it.id == member.streamId } ?: return null
        val chosen = choices[stream.id]?.takeIf { id -> stream.options.any { it.id == id } }
        val first = day.filter { index.member(it)?.streamId == stream.id }
            .minWithOrNull(compareBy({ timeKey(it.timeStart) }, { it.index }, { lessonKey(it) }))
        return Mark(stream.id, stream.options, chosen, first != null && lessonKey(first) == lessonKey(lesson))
    }

    internal fun teacherNames(raw: String): List<String> =
        raw.split(';').map { it.trim() }.filter { it.isNotEmpty() && it != "—" && it != "-" }

    internal fun teacherKey(name: String): String =
        name.trim().lowercase(Locale.forLanguageTag("ru")).replace('ё', 'е')
            .replace(Regex("[.\\s]+"), " ").trim()

    internal fun timeKey(raw: String): String {
        val match = Regex("""(\d{1,2}):(\d{2})""").find(raw) ?: return raw.trim()
        return "%02d:%s".format(match.groupValues[1].toInt(), match.groupValues[2])
    }

    private fun subjectKey(lesson: Lesson): String {
        val norm = lesson.subjectNormalized.trim()
        if (norm.isNotEmpty()) return norm
        return lesson.subjectRaw.trim().lowercase(Locale.forLanguageTag("ru")).replace('ё', 'е')
    }

    private fun lessonKey(lesson: Lesson): String = listOf(
        lesson.groupId, lesson.dayOfWeek, lesson.parity, lesson.index,
        timeKey(lesson.timeStart), subjectKey(lesson), teacherKey(lesson.teacherRaw), lesson.classroomRaw.trim()
    ).joinToString("|")

    private fun slotOptionId(lesson: Lesson): String {
        val teachers = teacherNames(lesson.teacherRaw).map { teacherKey(it) }.sorted().joinToString("+")
        return subjectKey(lesson) + "|" + teachers + "|" + lesson.classroomRaw.trim()
    }

    private data class Cluster(
        val streamId: String,
        val title: String,
        val sameSubject: Boolean,
        val lessons: List<Lesson>,
        val optionLabels: Map<String, String>
    )

    private class StreamBuilder(val id: String, val title: String, val sameSubject: Boolean) {
        var joined: Boolean = true
        val options = linkedMapOf<String, String>()
        fun build(): Stream = Stream(id, title, options.map { Option(it.key, it.value) }, joined && sameSubject)
    }

    private fun clusters(lessons: List<Lesson>): List<Cluster> {
        val found = mutableListOf<Cluster>()
        val seen = mutableSetOf<String>()
        for ((slot, rows) in lessons.groupBy { it.dayOfWeek to timeKey(it.timeStart) }) {
            for (week in weekCodes(rows)) {
                val active = rows.filter { it.parity == 0 || it.parity == week }
                val names = linkedMapOf<String, String>()
                for (lesson in active.sortedWith(compareBy({ it.index }, { subjectKey(it) }))) {
                    for (name in teacherNames(lesson.teacherRaw)) {
                        val id = teacherKey(name)
                        if (id.isNotEmpty()) names.putIfAbsent(id, name.trim())
                    }
                }
                if (names.size < 2 || active.isEmpty()) continue
                val subjects = active.map { subjectKey(it) }.filter { it.isNotEmpty() }.distinct()
                val same = subjects.size == 1
                val set = names.keys.sorted().joinToString("+")
                val streamId = if (same) "s:${subjects.first()}:$set" else "t:${slot.first}:${slot.second}:$set"
                val signature = streamId + "#" + active.map { lessonKey(it) }.sorted().joinToString(",")
                if (!seen.add(signature)) continue
                val labels = if (same) names else active.associate { slotOptionId(it) to slotLabel(it) }
                val title = active.minWith(compareBy({ it.dayOfWeek }, { timeKey(it.timeStart) }, { it.index })).subjectRaw
                found += Cluster(streamId, title, same, active, labels)
            }
        }
        return found
    }

    private fun slotLabel(lesson: Lesson): String {
        val teachers = teacherNames(lesson.teacherRaw).joinToString(", ")
        return listOf(lesson.subjectRaw.trim(), teachers).filter { it.isNotEmpty() }.joinToString(" · ")
            .ifBlank { teachers }
    }

    private fun weekCodes(rows: List<Lesson>): Set<Int> {
        val codes = rows.map { it.parity }.filter { it > 0 }.toSet()
        return if (codes.isEmpty()) setOf(1) else codes
    }
}
